### PR TITLE
Feat/os input

### PR DESCRIPTION
**Pull request by LucasLvy**

### CODE REVIEW

File: src/utils/definitions/general_config.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            chain_id: Felt252::new(0),
+            fee_token_address: Felt252::new(0),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct StarknetGeneralConfig {
+    starknet_os_config: StarknetOsConfig,
+    contract_storage_commitment_tree_height: usize,
```
Comment (LucasLvy) [inline src/utils/definitions/general_config.rs:40]: avoid using usize

File: src/utils/definitions/general_config.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+struct StarknetOsConfig {
+    chain_id: Felt252,
+    fee_token_address: Felt252,
+}
+
+// TODO: add real values
+impl Default for StarknetOsConfig {
+    fn default() -> Self {
+        StarknetOsConfig {
```
Comment (LucasLvy) [inline src/utils/definitions/general_config.rs:30]: Derive default directly and create another function to get the real values

File: src/utils/definitions/general_config.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    tx_commitment_tree_height: usize,
+    event_commitment_tree_height: usize,
+    cairo_resource_fee_weights: HashMap<String, f32>,
+    enforce_l1_handler_fee: bool,
+}
+
+pub static CONFIG: OnceCell<StarknetGeneralConfig> = OnceCell::const_new();
+
+impl Default for StarknetGeneralConfig {
+    fn default() -> Self {
```
Comment (LucasLvy) [inline src/utils/definitions/general_config.rs:57]: i don't think this should be the default value

File: src/utils/hasher/pedersen.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    fn hash_bytes(data: &[u8]) -> Felt252 {
+        // Calculate the number of 31-byte chunks we'll need, rounding up.
+        // (1 byte is used padding to prevent the value of field from being greater than the field's
+        // modulus) TODO: It is need a way to truncate bytes to fit into values smaller than modular
+        // (for optimization)
+        const CHUNK_SIZE: usize = 31;
+        let mut hash_value = Felt252::new(0);
+
+        for chunk in data.chunks(CHUNK_SIZE) {
+            // It is safe to unwrap here because we know that the chunk size is 31 and the value can not
```
Comment (LucasLvy) [inline src/utils/hasher/pedersen.rs:24]: there's no unwrap here lol

File: src/utils/hasher/pedersen.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ///
+    /// h(h(h(h(0, data\[0\]), data\[1\]), ...), data\[n-1\]), n).
+    #[inline]
+    fn compute_hash_on_elements(_elements: &[Felt252]) -> Felt252 {
+        // TODO: use impl from lambdaworks compute_hash_on_elements(elements)
+        Felt252::new(0)
+    }
+}
+
+// #[test]
```
Comment (LucasLvy) [inline src/utils/hasher/pedersen.rs:59]: remove commented code

File: src/utils/hasher/mod.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+use cairo_felt::Felt252;
+
+pub mod pedersen;
+
+/// A trait for hashing.
+pub trait HasherT {
+    /// Hashes the given data.
+    /// # Arguments
```
Comment (LucasLvy) [inline src/utils/hasher/mod.rs:8]: add line breaks and detail a bit how it's handled (31 bytes/31 bytes)

File: src/utils/commitment_tree/mod.rs [hunk 7]
```diff
@@ -1,3 +1,4 @@
 pub mod binary_fact_tree;
 pub mod leaf_fact;
+pub mod nodes;
```
Comment (LucasLvy) [inline src/utils/commitment_tree/mod.rs:3]: i think we should remove the whole commitment tree thing for now. we'll have a good implementation soon

File: src/storage/starknet.rs [hunk 8]
```diff
@@ -1,56 +1,60 @@
...
 #[derive(Serialize, Deserialize)]
-pub struct CommitmentInfo {
-    previous_root: FieldElement,
-    updated_root: FieldElement,
+pub struct CommitmentInfo<S: Storage, H: HasherT> {
+    pub previous_root: Felt252,
+    pub updated_root: Felt252,
     tree_height: usize,
     commitment_facts: CommitmentFacts,
+    _phantom_data: PhantomData<S>,
```
Comment (LucasLvy) [inline src/storage/starknet.rs:26]:
```suggestion
    _phantom_storage: PhantomData<S>,
```

File: src/storage/starknet.rs [hunk 9]
```diff
@@ -1,56 +1,60 @@
...
+    _phantom_data_2: PhantomData<H>,
```
Comment (LucasLvy) [inline src/storage/starknet.rs:27]:
```suggestion
    _phantom_hasher: PhantomData<H>,
```

File: src/fact_state/contract_state.rs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        // the outermost hash to guarantee unique "decoding".
+        H::hash_elements(hash_value, contract_state_hash_version)
+            .to_bytes_be()
+            .to_vec()
+    }
+}
+
+impl DBObject for ContractState {
+    fn db_key(suffix: Vec<u8>) -> Vec<u8> {
+        let prefix: &[u8] = "contract_state".as_bytes();
```
Comment (LucasLvy) [inline src/fact_state/contract_state.rs:56]:
```suggestion
        let prefix: &[u8] = b"contract_state";
```

File: src/fact_state/contract_state.rs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        let sep: &[u8] = ":".as_bytes();
```
Comment (LucasLvy) [inline src/fact_state/contract_state.rs:57]:
```suggestion
        let sep: &[u8] = b":";
```

File: src/error/mod.rs [hunk 12]
```diff
@@ -19,11 +19,11 @@ pub enum CommitmentInfoError {
     #[error("Inconsistent tree heights : {0} {1}.")]
     InconsistentTreeHeights(usize, usize),
     #[error("Inconsistent tree roots : {0} {1}.")]
-    InconsistentTreeRoots(FieldElement, FieldElement),
+    InconsistentTreeRoots(Felt252, Felt252),
```
Comment (LucasLvy) [inline src/error/mod.rs:22]: which is which ?

File: src/utils/definitions/general_config.rs (hunk 1 above)
Comment (EvolveArt) [reply src/utils/definitions/general_config.rs:40]: why ?

File: src/os_input.rs [hunk 13]
```diff
@@ -1,15 +1,24 @@
...
-    contract_class_commitment_info: CommitmentInfo,
-    class_hash_to_compiled_class_hash: HashMap<FieldElement, FieldElement>,
-    block_hash: FieldElement,
+struct StarknetOsInput<S: Storage, H: HasherT> {
+    contract_state_commitment_info: CommitmentInfo<S, H>,
+    contract_class_commitment_info: CommitmentInfo<S, H>,
+    deprecated_compiled_classes: HashMap<Felt252, Felt252>, // TODO: Add contract_class module
+    compiled_classes: HashMap<Felt252, Felt252>,            // TODO: Add contract_class module
+    contracts: HashMap<Felt252, ContractState>,
+    class_hash_to_compiled_class_hash: HashMap<Felt252, Felt252>,
```
[the remark this answers is not in the document]
Comment (LucasLvy) [reply src/os_input.rs:20]: class_hash_to_compiled_class_hash: HashMap<Felt252, Felt252>,

File: src/utils/definitions/general_config.rs (hunk 1 above)
Comment (LucasLvy) [reply src/utils/definitions/general_config.rs]: it's better to have a fixed size on every machine

File: src/os_input.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (EvolveArt) [reply src/os_input.rs:20]: What's wrong in it ?

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**